### ISSUE TITLE
autoforget_unhealthy for Tempo ingesters

### ISSUE THREAD

**Issue by chenfeilee** [grafana/tempo#1275]
Just wondering if the `autoforget_unhealthy` feature (implemented in Loki; see https://github.com/grafana/loki/pull/3919) has been made available for Tempo? We are deploying tempo distributed on ECS Fargate and it would be nice if we don't have to click the forget button manually every time we do a rolling update :)

**Comment by joe-elliott** [comment]
Currently only our compactors support autoforgetting unhealthy instances. This was the component our community was having the most with (#1081).

Does Fargate have a similar idea to "statefulset"? An ingester should come back up with the same name and access to the same disk which should prevent this issue from occurring. I'm guessing this doesn't quite work with fargate?

Even if this doesn't work an ingester that shuts down cleanly should remove itself from the ring and this shouldn't be an issue. Are you seeing unhealthy ingesters hanging out in the ring regularly?

**Comment by chenfeilee** [comment]
Thanks for the prompt response appreciate it!

> Does Fargate have a similar idea to "statefulset"?

Unfortunately no :( 

> Are you seeing unhealthy ingesters hanging out in the ring regularly?

Yes at the moment because we have not implemented graceful shutdown using the `shutdown` endpoint and queries are failling with `error finding ingesters in Querier.FindTraceByID: too many unhealthy instances in the ring`. Let me give it a try and see if the ingester remove itself from the ring after graceful shutdown. 

Nevertheless, this still feels like an amazing feature for certain use cases e.g. when graceful shutdown is not possible (ingester being OOM killed for example). Just curious if there is any plan to include this `autoforget_unhealthy` feature in Tempo any time soon? Seems like Cortex is trying to implement this too right now (https://github.com/cortexproject/cortex/issues/1521)

**Comment by joe-elliott** [comment]
So in the past we had discussed adding a config option to completely flush the ingesters on shutdown for situations like this. Currently you have to hit the `/flush` endpoint manually. This would help in this situation b/c your ingesters would automatically dump all data to the backend and remove themselves from the ring cleanly. Naturally this would require an increased shutdown grace period.

Auto forgetting ingesters should technically only be necessary if there are other issues going on with shutdown. A gracefully exited ingester should not remain in the ring.

Going to ping @mdisibio to see if he has any thoughts as well since he did the auto-forget compactor work.

**Comment by mdisibio** [comment]
Are you using RF=3 and rolling update 1 pod at a time?   These are required to prevent data loss when an ingester's local disk is deleted without calling `/flush`.

Regarding the two options: adding automatic `/flush` seems better because it avoids data loss under more scenarios (stateless ingesters as described here, but also other tasks like scaling down).  Auto-forget could look like a healthy cluster, but mask underlying data loss.  

Long-term I think both changes are useful.   Auto-forget seems simple on the surface, but requires switching the ring to the new `BasicLifecycler`, which may not be trivial.

**Comment by chenfeilee** [comment]
> Are you using RF=3 and rolling update 1 pod at a time

Not at the moment as we are still at the initial development phase but good point there

@joe-elliott I am slightly confused about the `/flush` vs `/shutdown` endpoints. Based on the documentation, seems like I don't have to call the `/flush` endpoint if I were to use the `/shutdown` endpoint?

**Comment by chenfeilee** [comment]
and also, when I hit the `/shutdown` endpoint, I do see this log line occasionally in the ingester shutting-down:  `Abandoning op in flush queue because ingester is shutting down`. Does this mean something undesirable?

**Comment by chenfeilee** [comment]
[log-shutdown-endpoint-only.txt](https://github.com/grafana/tempo/files/8045618/log-shutdown-endpoint-only.txt) is what I have retrieved from one of the ingester containers when shutting down. Doesn't seem like the container was deleted before shutdown could complete though. 

In the attached, I do see the 3 log messages you have pointed out above but I could also still see the line `Abandoning op in flush queue because ingester is shutting down` in there. 

I am running this on ECS and I have set `graceful_shutdown_timeout: 120s` in tempo config and also `stopTimeout : 120s` (Time duration (in seconds) to wait before the container is forcefully killed if it doesn't exit normally on its own) in the ECS task definition.

Also, seems like the `/shutdown` endpoint didn't manage to flush the traces to backend storage as I don't see the older traces in Grafana anymore (I am getting error `Status: 404 Not Found Body: trace not found`). 

Interestingly, I have tried calling the `/flush` endpoint before the `/shutdown` endpoint and looks like it helps (kindly refer to  
[log-both-flush-and-shutdown-endpoint.txt](https://github.com/grafana/tempo/files/8045712/log-both-flush-and-shutdown-endpoint.txt) for the log messages)
- I am seeing log lines (e.g. `"flushing block"`, `"object uploaded to s3" objectName=...`) from calling the `/flush` endpoint which tells me that the flushing did happen. These lines didn't appear when calling `/shutdown` endpoint only
- I am able to find the older traces in Grafana afterwards

**Comment by joe-elliott** [comment]
Yes, autoscaling on the ingesters is a bad idea :). We need to remove this option from the chart.

https://github.com/grafana/helm-charts/pull/1057#issuecomment-1111453689

Tempo's http endpoints are documented here:

https://grafana.com/docs/tempo/latest/api_docs/

and more extensive ring documentation can be found here:

https://grafana.com/docs/tempo/latest/operations/consistent_hash_ring/

**Comment by vaibhhavv** [comment]
Hi @joe-elliott , does `autoforget_unhealthy` released for rest of the tempo components? or do you have some ETA when we can expect this as a feature?

**Comment by joe-elliott** [comment]
Searching the codebase it looks like we've only added this to the metrics generator and compactor components. We would take a PR to extend this to the ingesters if someone is looking to contribute it.

**Comment by vaibhhavv** [comment]
@joe-elliott as you mentioned `autoforget_unhealthy` is added for compactor and metrics generator but I cannot find the same configurations in the docs [here](https://grafana.com/docs/tempo/latest/configuration/#compactor). Am I missing something?

**Comment by joe-elliott** [comment]
They are both set to the heartbeat config * 2:

https://github.com/grafana/tempo/blob/29a93af3f517849b51918b0076087d7d57408c97/modules/generator/generator.go#L109

https://github.com/grafana/tempo/blob/29a93af3f517849b51918b0076087d7d57408c97/modules/compactor/compactor.go#L76

**Comment by beanworks-eng-shared** [comment]
Is it possible to let the ring.NewAutoForgetDelegate to apply to the distributor and ingester? If not, is it possible to provide an endpoint to list and forget the unhealthy instance? 
Do you think we can remove it using the curl command? 
We found some instance crashes and the unhealthy instance not leave the ring.  and the instance has been terminated already.

**Comment by beanworks-eng-shared** [comment]
And we got this error: level=error ts=2024-09-19T16:36:41.337549603Z caller=rate_limited_logger.go:27 msg="pusher failed to consume trace data" err="at least 1 live replicas required, could only find 0 - unhealthy instances: X.X.X.X:9095"

**Comment by joe-elliott** [comment]
> Is it possible to let the ring.NewAutoForgetDelegate to apply to the distributor and ingester? 

I believe we briefly looked into this and there was something that made it more complicated than we had initially thought. Your ingesters ought to be relatively stable though. If this is a regular issue I would focus on ingester stability.

> If not, is it possible to provide an endpoint to list and forget the unhealthy instance?

Yes, the `/ingester/ring` [endpoint](https://grafana.com/docs/tempo/latest/api_docs/#ingesters-ring-status) on the distributor does this.

> Do you think we can remove it using the curl command?

Yes, but I don't know how off the top of my head. I believe a specifically crated POST to the ring endpoint will do it. I'd use the "network" tab in your browser to capture the request and replicate it with curl.

**Comment by joe-elliott** [comment]
Closing this issue b/c the symptom created by this issue (massive refused writes) goes away with our new upcoming architecture. This would have been nice but our attempts to move to the new ring did not work.

I do not think this will be important in the future.

**Comment by vaibhhavv** [comment]
Just to know @joe-elliott, what is the ETA of the above mentioned `new upcoming architecture`?

**Comment by joe-elliott** [comment]
Hoping for something available externally this fall-ish.
